Migration: Backend, Rake file, Photos import

tclaus opened the pull request:
This is just the backend part extracted from the larger Upload Migration PR #8274 
All UI stuff relies on this backend. 

Rake file for Command line Import is included. I hope that with this smaller PR (based on the latest development) we can let thinks work faster. 

@Flaburgan 
Salutations, mon ami.

tclaus committed:
Migration Backend Part

SuperTux88 commented on lib/archive_importer/post_importer.rb:
```diff
@@ -17,6 +17,7 @@ def substitute_author
 
       entity_data["photos"].each do |photo|
         photo["entity_data"]["author"] = user.diaspora_handle
+        photo["entity_data"]["remote_photo_path"] = "#{AppConfig.pod_uri}uploads\/images\/"
```
This only changes the photos URL if the post doesn't exist on the pod yet, changing it for posts which were already federated to this pod before the migration need to be changed separately.

Also this only changes the URLs on the pod where the user migrates to, all other pods which have received the posts from the old pod will still load the photos from there. So photos also need a way to update existing posts (both on the target pod and also on other remote pods), which hopefully can be done the same way. Sending a new photo message for each photo isn't a solution because that would just fill all sidekiq queues on all pods with photo messages. 

Also it also can't just blindly be set to the url of the new pod, because if the new pod uses something like S3, then the `remote_photo_path` would be the path to S3.

tclaus replied:
??? 
So.. how do a new photo URL can be send out to other pods to update their post?

tclaus replied:
As discussed elsewhere - this will be done elsewhere. (Protocol, Federation?)

SuperTux88 replied:
The first part of the comment is still valid and belongs to this PR:

> This only changes the photos URL if the post doesn't exist on the pod yet, changing it for posts which were already federated to this pod before the migration need to be changed separately.

It only changes the URL of photos which aren't in the database already. And during testing yesterday I also noticed, that it does this during the import of the JSON, so it then imports the photos with a broken URL (until I also import a photos archive). So this shouldn't happen here at all, it should happen during the photo import, and that way you also fix the problem that it ignores photos which are already in the database.

SuperTux88 commented on lib/tasks/accounts.rake:
```diff
@@ -2,43 +2,47 @@
 end
+
+def request_archive_parameter(args)
+  return unless args[:archive_path].nil?
```
I would create a generic `request_parameter` method with a `arg` (`:archive_path`) and `text` (`"Enter the archive (.json, .gz, .zip) path: "`) parameter (in addition to the `args` parameter), because otherwise all these method are very similar.

SuperTux88 commented on lib/archive_importer.rb:
```diff
@@ -36,8 +36,9 @@ def create_user(attr)
-    self.user = User.build(data)
-    user.save!
+    @user = User.find_or_build(data)
+    @user.getting_started = false
```
This is already set in the hash above, but it probably doesn't work because the `User.setup` method doesn't handle it. Instead of setting this separately here, I would just add it to the `setup` method, so the value from the hash works. This is basically the only "setting" I feel confident to always overwrite when migrating, because when a user imports an existing archive we can be sure it isn't a new user and doesn't need "getting started" anymore.

SuperTux88 reviewed (changes requested):
Thanks for the separate PR :cookie:

tclaus commented on app/models/photo.rb:
```diff
@@ -96,26 +96,24 @@ def processed?
+    # During migration in bad cases this might be happen
+    # If this happens, stream loading stops. Better dont show photos as stop loading stream
+    if remote_photo_path.present? && remote_photo_name.present?
+      name = "#{name.to_s}_" if name # rubocop:disable Lint/RedundantStringCoercion
```
'name' magically might be nil - and String interpolation will write "nil" but an empty string is needed.
I think for images in original size, without _small, _medium etc suffixes this might be the case.

SuperTux88 replied:
If `name` is `nil` then the `if name` at the end prevents this. But also "#{nil}" also creates an empty string, so it works the same as `.to_s`.

tclaus commented on app/models/user.rb:
```diff
@@ -312,6 +312,9 @@ def like_for(target)
   ######### Data export ##################
   mount_uploader :export, ExportedUser
 
+  ######### Photo export ##################
```
It felt better to see the mount_uploader: (mixins?) to be nearby. Then the processing parts follow in order.

SuperTux88 replied:
Before the `mount_uploader` were grouped together with the respective processing for them. The `User` model has a lot of comment header grouping together different parts and this mixes this up because now the `Photo export` header is above the processing for the `Data export`. If you want to make this one big group (and both the `mount_uploader` together at the top) then you would need to remove the second header and create a header for both data and photos together.

tclaus replied:
Makes sense.I'll change this back.

tclaus commented on app/models/user.rb:
```diff
@@ -404,23 +404,35 @@ def basic_profile_present?
   end
 
+  def self.find_or_build(opts={})
+    user = User.find_or_create_by(username: opts[:username], email: opts[:email])
```
This is (also) called from the rake file that has no user-context, but in command line a target username must be given. Mail-Address then is read from the import file. 

Maybe this behavior should be well defined and then fixed or documented somewhere.

SuperTux88 replied:
I'm fine with reading the mail-address from the archive, but it shouldn't overwrite the existing mail-address and especially it should still work if the mail-addresses are different. So basically only set the mail-address if it's otherwise missing/empty.

Flaburgan replied:
Related, the import currently fails if the e-mail in the archive is already taken by another account in the pod. What do we want to do with that?

tclaus replied:
I've already changed this. Only in case a migration is started from a rake file. Its now searched only for the username. In case a username was not found, one is created with attached email.
In case of an online migration, this code is not called, user exists anyway. Username and mail provided in the archive is ignored. 

But what if - a user must be created and the email address is already occupied?

Flaburgan replied:
>In case a username was not found, one is created with attached email

By attached email you mean the one in the archive right? This is the problem here, it will fail if another user already uses that email, as the  same e-mail can't be used twice.

SuperTux88 replied:
There is not much more you can do about it than just show the error. If a user already has an account with the same mail-address on that pod, then they either need to import it into their existing user, or create a second account with a different mail-address. It's the same as if you just register normally, if the user wants to register with the same mail-address twice it fails.

tclaus replied:
Process fails fast in this case.

Flaburgan replied:
>There is not much more you can do about it than just show the error

Well, that's the question. Should we block the users who's trying to import the archive saying that this e-mail address is already taken (which means they either have to log in the other account and modify the address, or modify the JSON, which is a pain, we don't want that) or should we just ignore the value of the email in the archive and do the import anyway? I honestly don't know.

SuperTux88 replied:
If the user imports the archive into an existing account, the mail-address is already ignored (at least it should be). It only fails if you import the archive using the rake task into a non-existing account, then creation of the account fails, because there is already another account with the same mail-address. So either use the rake-task to import it into the other account existing account or create a new account first with a different email-address.

tclaus replied:
Exactly.

tclaus commented on app/models/user.rb:
```diff
@@ -404,23 +404,35 @@ def basic_profile_present?
     self.color_theme ||= AppConfig.settings.default_color_theme
-    self.valid?
+    self.strip_exif = opts.fetch(:strip_exif, true)
+    self.show_community_spotlight_in_stream = opts.fetch(:show_community_spotlight_in_stream, false)
```
Moved it to the importer

tclaus commented on app/uploaders/exported_photos.rb:
```diff
@@ -9,7 +9,14 @@ def store_dir
   def filename
-    "#{model.username}_photos_#{secure_token}.zip" if original_filename.present?
+    if original_filename.present? # rubocop:disable Style/GuardClause
+      extension = File.extname(@filename) if @filename
```
I'll test it. And also for exported_user.rb

tclaus commented on lib/tasks/accounts.rake:
```diff
@@ -2,43 +2,47 @@
-      puts "Migration record already exists for the user, can't continue"
+    start_time = Time.now.getlocal
+    if args[:new_user_name].present?
+      import_profile = ImportProfileService.new
```
Added.

tclaus commented on app/uploaders/exported_user.rb:
```diff
@@ -10,10 +10,13 @@ def store_dir
-    "#{model.username}_diaspora_data_#{secure_token}.json.gz" if original_filename.present?
+    if original_filename.present? # rubocop:disable Style/GuardClause
+      extension = File.extname(@filename) if @filename
+      "#{model.username}_data_#{secure_token}#{extension}"
```
Its for down- and uploads. A user also might upload an uncompressed json. 
This will rescue the multi-point suffice
```suggestion
    return if original_filename.blank?

    filename_parts = original_filename.split(".")
    filename_parts.slice!(0)
    extensions = filename_parts.join(".")
    "#{model.username}_data_#{secure_token}.#{extensions}"
```

tclaus commented on lib/archive_importer.rb:
```diff
@@ -60,14 +61,14 @@ def set_auto_follow_back_aspect
 
-    aspect = user.aspects.find_by(name: name)
-    user.update(auto_follow_back_aspect: aspect) if aspect
+    aspect = @user.aspects.find_by(name: name)
```
The usage of at (@) is the default for using accessors. There are problems writing to `user`, but accessing `@user` from outside. Rubocop is also whining.

tclaus replied:
The warning was: RuboCop: Redundant `self` detected. [Style/RedundantSelf]
besides of that, I've tried to unify the self.user, user and @user notations in this file. The at-sign is IMHO the most clear statement that an instance variable is addressed. So I changed it to this notation. 
You may be right - "user" accesses the accessor, but the '@' makes it more clear what happens. I think, code style should rule that class variables should be used with the '@' - regardless if there is an accessor or not.

SuperTux88 replied:
We usually use the accessors (where they are available) and not the variable directly (maybe there are a few places where variables were access directly, but some have a special reason and others maybe are inconsistency from the past). If don't want to use the accessor and only use the variable directly, then we wouldn't even need to define the acccessor. Since the old code didn't have any rubocop warnings, I don't know why we would need to change it.

tclaus replied:
If this is common 'diaspora' style, I won't like to change it.

tclaus commented:
@Flaburgan @SuperTux88 I've sent  some changes. What else is a MUST to change here?

tclaus commented on app/models/photo.rb:
```diff
@@ -96,26 +96,24 @@ def processed?
-    if remote_photo_path
-      name = name.to_s + '_' if name
+  def url(name=nil)
+    # During migration in bad cases this might be happen
```
Hm - I think I discovered this once. Leave it or remove it?

SuperTux88 replied:
You can leave the double checks, since I don't think it harms anything ... but this comment confuses me more than it helps. It doesn't say "what" maybe happens (which maybe is less a problem with the context of this diff, but if you look at the code only, I have no idea what the problem should be. So I guess you mean `remote_photo_name` could be nil while `remote_photo_path` is set? But that should never happen, they should always be set together, only setting one makes no sense. But also `remote_photo_name` should also never be nil in an import anyway, since it's a required field? So I say, just remove the comment, the code is fine as it was but I also think it's fine with the second check now.

tclaus commented on app/workers/import_profile.rb:
```diff
@@ -0,0 +1,24 @@
+    include Diaspora::Logging
+
+    def perform(user_id)
+      user = User.find_by(username: user_id)
```
Changed to user.id - is the check really unneeded?

SuperTux88 replied:
If the user existed when the job was created, it will exist when it is executed, since diaspora never deletes users. So I don't see why the check would be needed. The only reason this could fail would be if a podmin manually does things in the database, and when podmins do that, all sort of things break, and then it can just fail hard.

tclaus commented on spec/integration/migration_service_spec.rb:
```diff
@@ -271,10 +271,11 @@ def setup_validation_time_expectations
       ).to match_array(
         status_message_with_photos_entity.photos.map {|photo|
-          [photo.guid, photo.text, photo.remote_photo_path, photo.remote_photo_name, photo.width, photo.height]
+          # remote_photo_path is rewritten with new local path
```
This test imports an archive, not the photos. Should the photos path be rewritten before the photos are actually imported?

Flaburgan replied:
If I read https://github.com/diaspora/diaspora/pull/8298#discussion_r719851399 correctly I would say no.

SuperTux88 replied:
No it shouldn't, so this test should just be reverted how it was before.

tclaus commented on app/services/import_profile_service.rb:
```diff
@@ -0,0 +1,97 @@
+        next unless File.exist?(uploaded_file) && photo.remote_photo_name.present?
+
+        File.open(uploaded_file) do |file|
+          photo.random_string = File.basename(uploaded_file, ".*")
```
Puh.. Maybe the right and reliable handling of photos should be done in another PR?

SuperTux88 replied:
Fine with me, photos are definitely a bigger topic. and it also needs to be split up into two PRs anyway to completely do everything, since parts also need to be backported to 0.7.x

tclaus commented on app/uploaders/exported_user.rb:
```diff
@@ -10,10 +10,15 @@ def store_dir
+    return if original_filename.blank?
+
+    filename_parts = original_filename.split(".")
+    filename_parts.slice!(0)
+    extensions = filename_parts.join(".")
```
What I want to do is extend the filename with a 'secure' part and attach the extension(s). 
bob.json.gz => bob_wjhcxbwhjxcbw.json.gz

diaspora-code-review commented on app/uploaders/secure_uploader.rb:
```diff
@@ -1,8 +1,19 @@
 # frozen_string_literal: true
 
 class SecureUploader < CarrierWave::Uploader::Base
+
```
Layout/EmptyLinesAroundClassBody: Extra empty line detected at class body beginning.

diaspora-code-review commented on lib/archive_importer.rb:
```diff
@@ -60,7 +62,7 @@ def set_auto_follow_back_aspect
     return if name.nil?
 
-    aspect = user.aspects.find_by(name: name)
+    aspect = self.user.aspects.find_by(name: name)
```
Style/RedundantSelf: Redundant `self` detected.

SuperTux88 reviewed (changes requested):
So there is still the security problem with importing photos. But I'm fine with excluding photos import from this PR, but then it needs to be removed from here, since the current state is an active security problem.

I've also created https://github.com/diaspora/diaspora_federation/pull/119 for the federation part of the photos-migration, to tell other pods the new path so they can migrate photos on bulk and don't need a message for every single photo. If there are conflicts with filenames, where a photo from the archive can't be imported, it either needs to be deleted (and then also federate the deletion), or be imported on a different filename (and then it would need to federate the updated filename with a single photo message). But since I don't expect many conflicts, this should be fine, but it still needs to be handled correctly.

SuperTux88 reviewed (changes requested):
The PR still contains the photo import logic including the security problem, if you want to move the handling of photos to another PR (which I'm fine with), it needs to be removed here.

And besides a few small other things I think this is looking good. :+1:

msph31 commented on lib/archive_importer.rb:
```diff
@@ -31,12 +31,14 @@ def create_user(attr)
     )
-    self.user = User.build(data)
+    self.user = User.find_or_build(data)
+    user.show_community_spotlight_in_stream = data.fetch(:show_community_spotlight_in_stream, true)
```
you can overwrite it, as long as you dont overwrite existing posts/comments/conversations
(settings is less critical)

SuperTux88 committed:
Move extension logic to SecureUploader class

SuperTux88 committed:
Refactory archive concurrency so the same logic can be reused

SuperTux88 committed:
Allow to choose to overwrite settings and profile data

tclaus commented on app/workers/import_user.rb:
```diff
@@ -0,0 +1,36 @@
+    include Diaspora::Logging
+
+    def perform(user_id)
+      if currently_running_exports >= AppConfig.settings.export_concurrency.to_i
```
I am fine with changing setting to 'archive_jobs_concurrency' and changed it. For the shared code I am now using a base class for the archive.

SuperTux88 committed:
Send new remote_photo_path in migration message

tclaus commented:
Fixes #8255

SuperTux88 committed:
Don't overwrite photos of other users during import

If a photo with the same filename already exists, generate a new random
filename, and re-federate the photo with that filename. This ensures
users can't modify their archive to overwrite other users photos.

tclaus commented:
Looks good to me ! 👍🏻
Thanks to Benjamin for the hard rework.

SuperTux88 commented:
Merged, thanks for the initial work and for the fast review of my changes to it @tclaus :cookie:
